includes/class-wc-payments-dependency-service.php

```diff
@@ -0,0 +1,330 @@
+<?php
+/**
+ * WC_Payments_Dependency_Service class
+ *
+ * @package WooCommerce\Payments
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+use WCPay\Exceptions\Invalid_Dependency_Exception;
+
+/**
+ * Validates dependencies (core, plugins, versions) for WCPAY
+ * Used in the plugin main class for validation.
+ */
+class WC_Payments_Dependency_Service {
+
+	/**
+	 * Cached result of dependency check.
+	 *
+	 * @var bool | null
+	 */
+	private $has_valid_dependencies;
+
+	/**
+	 * Cached result of the first invalid dependency.
+	 *
+	 * @var string | null
+	 */
+	private $invalid_dependency;
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+
+		add_filter( 'admin_notices', [ $this, 'display_admin_notices' ] );
+	}
+
+	/**
+	 * Checks if all the dependencies needed to run WooCommerce Payments are present
+	 *
+	 * @return bool True if all required dependencies are met.
+	 */
+	public function has_valid_dependencies() {
+
+		if ( defined( 'WCPAY_TEST_ENV' ) && WCPAY_TEST_ENV ) {
+			return true;
+		}
+
+		// Return cached result, if available.
+		if ( is_bool( $this->has_valid_dependencies ) ) {
+			return $this->has_valid_dependencies;
+		}
+
+		$passed = true;
+
+		try {
+
+			$this->is_woo_core_active();
+			$this->is_woo_core_version_compatible();
+			$this->is_wc_admin_enabled();
+			$this->is_wc_admin_version_compatible();
+			$this->is_wp_version_compatible();
+		} catch ( Invalid_Dependency_Exception $e ) {
+			$passed = false;
+
+			// save invalid dependency to use on admin notice.
+			$this->invalid_dependency = $e->getMessage();
+		} finally {
+			// cache result before returning it.
+			$this->has_valid_dependencies = $passed;
+
+			return $passed;
+		}
+	}
+
+	/**
+	 * Render admin notices for unmet dependencies. Called on the admin_notices hook.
+	 *
+	 * @return null.
+	 */
+	public function display_admin_notices() {
+
+		// Do not show alerts while installing plugins.
+		if ( self::is_at_plugin_install_page() ) {
+			return;
+		}
+
+		if ( false === $this->has_valid_dependencies() ) {
+			WC_Payments::display_admin_error( $this->get_notice_for_exception_code( $this->invalid_dependency ) );
+		}
+	}
+
+	/**
+	 * Checks if WooCommerce is installed and activated.
+	 *
+	 * @return bool True if WooCommerce is installed and activated.
+	 *
+	 * @throws Invalid_Dependency_Exception If this dependency check does not pass.
+	 */
+	public function is_woo_core_active() {
+		// Check if WooCommerce is installed and active.
+		if ( ! class_exists( 'WooCommerce' ) ) {
+			throw new Invalid_Dependency_Exception( Invalid_Dependency_Exception::WOOCORE_NOT_FOUND );
+		}
+
+		return true;
+	}
```

Comment:
I like the new approach more than the previous one, but it's a little bit weird to see functions all either a) throw an Exception; or b) return `true`.

Maybe we should just change the functions to be called `throw_if_xxx()` and not return anything? Alternatively we could `return false` instead of throwing and then set the appropriate message based on the result from each function?

I think returning `false` would make the functions more usable throughout the codebase, but I'm not even sure if it makes sense to use these functions in other parts of the codebase 🤔 .

Reply:
> I think returning false would make the functions more usable throughout the codebase, but I'm not even sure if it makes sense to use these functions in other parts of the codebase 🤔 .

It's not used now but it might be in the future. That's why I didn't leave the methods private. 
I used the exceptions because I didn't want to add a bunch of if statements in the ```has_valid_dependencies```, but you're right, the way it's now can cause confusion. I'll rewrite it and get rid of the exceptions.

Reply:
@reykjalin , I had success with WC Admin 0.23.1 and WooCommerce 4.4. WC Admin 0.23.1 won't work with the latest WooCommerce version.